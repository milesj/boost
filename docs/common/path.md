# `Path`

A `Path` class is an immutable abstraction around file/module paths and the Node.js `fs` and `path`
modules. It aims to solve cross platform and operating system related issues in a straight forward
way. To begin, import and instantiate the `Path` class, with either a single path, or a list of path
parts that will be joined.

```ts
import { Path } from '@boost/common';

const absPath = new Path('/root/some/path');
const relPath = new Path('some/path', '../move/around', 'again');
```

The following methods are available on the class instance. The `PortablePath` type is a union of
`string | Path`.

- `append(...parts: PortablePath[]): Path` - Append path parts to the end of the current path and
  return a new `Path` instance.
- `equals(path: PortablePath): boolean` - Returns true if both paths are equal using strict
  equality.
- `ext(withoutPeriod?: boolean): string` - Return the extension (if applicable) with or without
  leading period.
- `exists(): boolean` - Return true if the current path exists on the file system.
- `isAbsolute(): boolean` - Return true if the current path is absolute.
- `isDirectory(): boolean` - Return true if the current path is a folder.
- `isFile(): boolean` - Return true if the current path is a file.
- `name(withoutExtension?: boolean): string` - Return the file name (with optional extension) or
  folder name.
- `parent(): Path` - Return the parent folder as a new `Path` instance.
- `path(): FilePath` - Return the current path as a normalized string.
- `prepend(...parts: PortablePath[]): Path` - Prepend path parts to the beginning of the current
  path and return a new `Path` instance.
- `relativeTo(to: PortablePath): Path` - Return a new relative `Path` instance from the current
  "from" path to the defined "to" path.
- `toString(): FilePath` - Return the current path as a normalized string.

## Resolved Paths

By default, the `Path` class operates on the defined path parts as-is. If you would prefer to
operate against real or resolved paths, use the `Path#realPath()` and `Path#resolve()` methods
respectively. The current path is
[resolved against](https://nodejs.org/api/path.html#path_path_resolve_paths) the defined current
working directory (`process.cwd()`).

```ts
path.path(); // Possibly inaccurate
path.resolve().path(); // Resolved accurately
```

## Static Factories

The static `Path.create()` and `Path.resolve()` methods can be used to factory a `Path` instance
from a string or an existing instance. Especially useful when used in combination with the
`PortablePath` type.

```ts
Path.create('some/file/path'); // Path
```
