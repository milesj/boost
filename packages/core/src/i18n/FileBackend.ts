/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import fs from 'fs-extra';
import path from 'path';
import i18next from 'i18next';

export interface FileBackendOptions {
  resourcePaths: string[];
}

export default class FileBackend {
  fileCache: { [path: string]: i18next.ResourceKey } = {};

  options: FileBackendOptions = {
    resourcePaths: [],
  };

  services: any = {};

  type: string = 'backend';

  init(services: any, options: Partial<FileBackendOptions>) {
    this.services = services;
    this.options = { ...this.options, ...options };
  }

  read(
    language: string,
    namespace: string,
    callback: (error: Error | null, resources: any) => void,
  ): i18next.ResourceKey {
    let resources = {};

    this.options.resourcePaths.forEach(resourcePath => {
      const filePath = this.services.interpolator.interpolate(
        path.join(resourcePath, '{{language}}/{{namespace}}.json'),
        {
          language,
          namespace,
        },
      );
      let contents = {};

      if (this.fileCache[filePath]) {
        contents = this.fileCache[filePath];
      } else if (fs.existsSync(filePath)) {
        contents = fs.readJsonSync(filePath);
        this.fileCache[filePath] = contents;
      }

      resources = {
        ...resources,
        ...contents,
      };
    });

    callback(null, resources);

    return resources;
  }
}
