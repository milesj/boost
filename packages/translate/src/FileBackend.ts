import fs from 'fs';
import path from 'path';
import i18next from 'i18next';
import { parseFile, Contract, Path, Predicates } from '@boost/common';
import { Locale, Format } from './types';

const EXTS: { [K in Format]: string[] } = {
  js: ['js'],
  json: ['json5', 'json'],
  yaml: ['yaml', 'yml'],
};

export interface FileBackendOptions {
  format?: Format;
  paths?: Path[];
}

export default class FileBackend extends Contract<FileBackendOptions>
  implements i18next.BackendModule {
  fileCache: { [path: string]: i18next.ResourceKey } = {};

  type: 'backend' = 'backend';

  init(services: unknown, options: Partial<FileBackendOptions>) {
    this.configure(options);

    // Validate resource paths are directories
    this.options.paths.forEach(resourcePath => {
      if (fs.existsSync(resourcePath) && !fs.statSync(resourcePath).isDirectory()) {
        throw new Error(`Resource path "${resourcePath}" must be a directory.`);
      }
    });
  }

  blueprint({ array, string }: Predicates) /* infer */ {
    return {
      format: string('yaml').oneOf(['js', 'json', 'yaml']),
      paths: array(string()),
    };
  }

  create() {
    // We don't need this but is required by the interface
  }

  read(
    locale: Locale,
    namespace: string,
    callback: (error: Error | null, resources: i18next.ResourceLanguage) => void,
  ): i18next.ResourceKey {
    const { format, paths } = this.options;
    const resources: i18next.ResourceKey = {};

    paths.forEach(resourcePath => {
      EXTS[format].some(ext => {
        const filePath = path.normalize(path.join(resourcePath, locale, `${namespace}.${ext}`));

        if (!this.fileCache[filePath] && !fs.existsSync(filePath)) {
          return false;
        }

        Object.assign(
          resources,
          this.fileCache[filePath] || (this.fileCache[filePath] = parseFile(filePath)),
        );

        return true;
      });
    });

    callback(null, resources);

    return resources;
  }
}
