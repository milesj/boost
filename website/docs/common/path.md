---
title: Path
---

A `Path` class is an immutable abstraction around file/module paths and the Node.js `fs` and `path`
modules. It aims to solve cross platform and operating system related issues in a straight forward
way. To begin, import and instantiate the `Path` class, with either a single path, or a list of path
parts that will be joined.

```ts
import { Path } from '@boost/common';

const absPath = new Path('/root/some/path');
const relPath = new Path('some/path', '../move/around', 'again');
```

## Resolved paths

By default, the `Path` class operates on the defined path parts as-is. If you would prefer to
operate against real or resolved paths, use the `Path#realPath()` and `Path#resolve()` methods
respectively. The current path is
[resolved against](https://nodejs.org/api/path.html#path_path_resolve_paths) the defined current
working directory (`process.cwd()`).

```ts
path.path(); // Possibly inaccurate
path.resolve().path(); // Resolved accurately
```

## Static factories

The static `Path.create()` and `Path.resolve()` methods can be used to factory a `Path` instance
from a string or an existing instance. Especially useful when used in combination with the
`PortablePath` type.

```ts
Path.create('some/file/path'); // Path
```

## API

The following methods are available on the class instance. In the examples below, the `PortablePath`
type is a union of `FilePath | Path`, and `FilePath` is a `string`.

### `.create`

> Path.create(filePath: PortablePath): Path

Create and return a new `Path` instance if a string. If already a `Path`, return as is.

```ts
const a = Path.create('foo');
const b = Path.create(new Path('bar'));
```

### `.resolve`

> Path.resolve(filePath: PortablePath, cwd?: PortablePath): Path

Like `create()` but also resolves the path against the current working directory (defaults to
`process.cwd()`).

```ts
const path = Path.resolve('foo', '/cwd'); // => /cwd/foo
```

### `append`

> Path#append(...parts: PortablePath[]): Path

Append path parts to the end of the current path and return a new `Path` instance.

```ts
const a = new Path('foo', 'bar'); // => foo/bar
const b = a.append('baz'); // => foo/bar/baz
```

### `equals`

> Path#equals(path: PortablePath): boolean

Returns true if both paths are equal using strict equality.

```ts
const a = new Path('foo');
const b = new Path('bar');

a.equals(b); // => false
```

### `ext`

> Path#ext(withoutPeriod?: boolean): string

Return the extension (if applicable) with or without leading period.

```ts
const path = new Path('some/path/to/file.js');

path.ext(); // => .js
path.ext(true); // => js
```

### `exists`

> Path#exists(): boolean

Return true if the current path exists on the file system.

```ts
const path = new Path('some/path/to/file.js');

path.exists(); // => false
```

### `isAbsolute`

> Path#isAbsolute(): boolean

Return true if the current path is absolute. _Does not_ verify existence on the file system.

```ts
const nixPath = new Path('/some/path/to/file.js');
const winPath = new Path('C:/some/path/to/file.js');

nixPath.exists(); // => true
winPath.exists(); // => true
```

### `isDirectory`

> Path#isDirectory(): boolean

Return true if the current path is a folder.

```ts
const filePath = new Path('some/path/to/file.js');
const folderPath = new Path('some/path/to/folder');

filePath.isDirectory(); // => false
folderPath.isDirectory(); // => true
```

### `isFile`

> Path#isFile(): boolean

Return true if the current path is a file.

```ts
const filePath = new Path('some/path/to/file.js');
const folderPath = new Path('some/path/to/folder');

filePath.isFile(); // => true
folderPath.isFile(); // => false
```

### `name`

> Path#name(withoutExtension?: boolean): string

Return the file name (with optional extension) or folder name.

```ts
const path = new Path('some/path/to/file.js');

path.name(); // => file.js
path.name(true); // => file
```

### `parent`

> Path#parent(): Path

Return the parent folder as a new `Path` instance.

```ts
const path = new Path('foo/bar/baz');
const parent = path.parent(); // => foo/bar
```

### `path`

> Path#path(): FilePath

Return the current path as a normalized string. Is also triggered when the `Path` instance is cast
to a string.

```ts
const path = new Path('foo/bar/baz');

path.path(); // => foo/bar/bar
```

### `prepend`

> Path#prepend(...parts: PortablePath[]): Path

Prepend path parts to the beginning of the current path and return a new `Path` instance.

```ts
const a = new Path('foo', 'bar'); // => foo/bar
const b = a.append('baz'); // => bar/foo/bar
```

### `relativeTo`

> Path#relativeTo(to: PortablePath): Path

Return a new relative `Path` instance from the current "from" path to the defined "to" path.

```ts
const from = new Path('/foo/bar/baz');
const to = from.relativeTo('/foo/qux'); // => ../../qux
```

### `toString`

> Path#toString(): FilePath

Return the current path as a normalized string. See [path()](#path).
