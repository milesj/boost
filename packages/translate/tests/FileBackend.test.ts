import path from 'path';
import FileBackend from '../src/FileBackend';

describe('FileBackend', () => {
  let backend: FileBackend;

  beforeEach(() => {
    backend = new FileBackend({
      paths: [path.join(__dirname, 'res')],
    });
  });

  describe('init()', () => {
    it('sets options on class', () => {
      backend.init({}, { format: 'yaml' });

      expect(backend.options.format).toBe('yaml');
    });

    it('errors if resource path is not a directory', () => {
      expect(() => {
        backend.init(
          {},
          {
            paths: [path.join(__dirname, '../package.json')],
          },
        );
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('read()', () => {
    it('supports .js extension', () => {
      backend.options.format = 'js';

      expect(backend.read('en', 'type-js', () => {})).toEqual({ type: 'js' });
    });

    // it('supports .mjs extension', () => {
    //   backend.options.format = 'js';

    //   expect(backend.read('en', 'type-mjs', () => {})).toEqual({ type: 'mjs' });
    // });

    it('supports .json extension', () => {
      backend.options.format = 'json';

      expect(backend.read('en', 'type-json', () => {})).toEqual({ type: 'json' });
    });

    it('supports .json5 extension', () => {
      backend.options.format = 'json';

      expect(backend.read('en', 'type-json5', () => {})).toEqual({ type: 'json5' });
    });

    it('supports .yaml extension', () => {
      backend.options.format = 'yaml';

      expect(backend.read('en', 'type-yaml', () => {})).toEqual({ type: 'yaml' });
    });

    it('supports .yml extension', () => {
      backend.options.format = 'yaml';

      expect(backend.read('en', 'type-yaml-short', () => {})).toEqual({
        type: 'yaml',
        short: true,
      });
    });

    it('returns empty object for missing locale', () => {
      expect(backend.read('fr', 'unknown', () => {})).toEqual({});
    });

    it('returns object for defined locale', () => {
      expect(backend.read('en', 'common', () => {})).toEqual({ key: 'value' });
    });

    it('caches files after lookup', () => {
      const key = path.normalize(path.join(__dirname, 'res/en/common.json'));

      expect(backend.fileCache[key]).toBeUndefined();

      backend.read('en', 'common', () => {});

      expect(backend.fileCache[key]).toBeDefined();
      expect(backend.fileCache[key]).toEqual({ key: 'value' });
    });

    it('passes the resources to the callback', () => {
      const spy = jest.fn();

      backend.read('en', 'common', spy);

      expect(spy).toHaveBeenCalledWith(null, { common: { key: 'value' } });
    });

    it('merges objects from multiple resource paths', () => {
      backend.options.paths.push(path.join(__dirname, 'res-more'));

      expect(backend.read('en', 'common', () => {})).toEqual({ key: 'value', more: true });
    });
  });
});
