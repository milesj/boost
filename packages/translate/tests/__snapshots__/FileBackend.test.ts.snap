// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileBackend init() errors if resource path is not a directory 1`] = `"Resource path \\"/Users/Miles/Sites/boost/packages/translate/package.json\\" must be a directory."`;
